src/twisted/web/http.py

```diff
@@ -1063,29 +1021,45 @@ def requestReceived(self, command, path, version):
             self.path, argstring = x
             self.args = parse_qs(argstring, 1)
 
-        # Argument processing
-        args = self.args
-        ctype = self.requestHeaders.getRawHeaders(b"Content-Type")
-        if ctype is not None:
-            ctype = ctype[0]
+        # Parse form requests into self.args
+        # TODO: Make parsing optional or only do that on-demand
 
+        ctype = self.requestHeaders.getRawHeaders(b"Content-Type")
         if self.method == b"POST" and ctype and clength:
-            mfd = b"multipart/form-data"
-            key = _parseContentType(ctype)
-            if key == b"application/x-www-form-urlencoded":
-                args.update(parse_qs(self.content.read(), 1))
-            elif key == mfd:
+            try:
+                key, opts = multipart.parse_options_header(str(ctype[0], "ASCII"))
+            except UnicodeDecodeError:
+                # Content-Type and all options must be ASCII
+                self.channel._respondToBadRequestAndDisconnect()
+                return
+
+            if key == "application/x-www-form-urlencoded":
+                self.args.update(parse_qs(self.content.read(), 1))
+                self.content.seek(0, 0)
+            elif key == "multipart/form-data" and "boundary" in opts:
                 try:
-                    self.content.seek(0)
-                    content = self.content.read()
-                    self.args.update(_getMultiPartArgs(content, ctype))
-                except _MultiPartParseException:
-                    # It was a bad request.
+                    parser = multipart.MultipartParser(
+                        stream=self.content,
+                        boundary=opts["boundary"],
+                        charset="charmap",
+                        # TODO: Make limits configurable. The default limits are
+                        # very generous to not break existing apps that rely on
+                        # large form fields.
+                        part_limit=1024,
```

Comment:
We can get these values via instance members of `t.web.http.Request` ... maybe start with exposing/documenting `part_limit` and have spool and memory limit as multiple of `part_limit`